src/shared/firewall-util-nft.c

```diff
@@ -0,0 +1,853 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <arpa/inet.h>
+#include <endian.h>
+#include <errno.h>
+#include <stddef.h>
+#include <string.h>
+#include <linux/netfilter/nf_nat.h>
+#include <linux/netfilter/nf_tables.h>
+#include <linux/netfilter_ipv4.h>
+#include <netinet/ip.h>
+#include <netinet/udp.h>
+
+#include "sd-netlink.h"
+
+#include "alloc-util.h"
+#include "firewall-util.h"
+#include "in-addr-util.h"
+#include "macro.h"
+#include "socket-util.h"
+
+#include <stdio.h>
+
+static const char nft_systemd_dnat_map_name[] = "map_port_ipport";
+static const char nft_systemd_table_name[] = "io.systemd.nat";
+static sd_netlink *nfnl;
```

Comment:
global static variables is something we really don't like, because of threading and so on. Please find a different place to store this, maybe pass in a generic context object or so to these functions, where you can store this

Reply:
What about something like this:
src/shared/firewall-util.h:

typedef struct fw_util_src ctx fw_util_ctx;
int fw_util_ctx_new(fw_util_ctx **);
 int fw_add_masquerade(
+               fw_util_ctx *ctx,
                 bool add,
There would be a firewall-util-private.h with
struct fw_util_ctx {
       union {
               /* backend specific .... */

The nft backend would then keep the nfnetlink handle in this union.
The object would have to be managed by nspawn/networkd.

I think if this route is taken it would make sense to open a new pull request that only has the fw-util API rework without
any of the nftables changes. What do you think? Any better idea?

Reply:
I've added a fw_ctx that gets passed around; it now contains what was formerly kept in 'static' variable. Please clarify/advise if thats not what you had in mind.